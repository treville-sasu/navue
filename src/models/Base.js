export class Model {
  // eslint-disable-next-line no-unused-vars
  constructor({ type, ...properties } = {}) {
    Object.assign(this, properties);
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  // TODO write specs
  clone() {
    return this.constructor.from(JSON.parse(JSON.stringify(this)));
  }

  static from({ type, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) return new this(properties);
    else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}

// TODO : implementation for when Vuejs Suppots Array-like objects.
// export class Store extends Array {
export class Store {
  constructor(properties = {}, ...items) {
    this.items = items;
    Object.assign(this, properties);
  }

  // get items() {
  //   return [...this];
  // }

  [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  get length() {
    return this.items.length;
  }

  valueOf() {
    return this.items;
  }

  get properties() {
    // eslint-disable-next-line no-unused-vars
    let { items, ...props } = { ...this };
    return props;
    // return Object.fromEntries(
    //   Object.entries(this).filter(([key]) => isNaN(key))
    // );
  }

  get first() {
    return this.items[0];
  }

  get last() {
    return this.items[this.length - 1];
  }

  add(item, index) {
    index ? this.items.splice(index, 0, item) : this.items.push(item);
    return this.length;
  }

  remove(item, index) {
    if (index === undefined) index = this.items.indexOf(item);
    return index >= 0 ? this.items.splice(index, 1).length == 1 : false;
  }

  keep(filter) {
    let i = this.items.length;
    while (i--) {
      let result = filter.call(this.items, this.items[i]);

      if (result == undefined) this.items.splice(i, 1);
      else if (!Object.is(result, this.items[i]))
        this.items.splice(i, 1, result);
    }
    return this;
  }

  toJSON() {
    return {
      ...this.properties,
      items: this.items,
      type: this.constructor.name
    };
  }

  static from({ type, items = [], ...properties }, ...constructors) {
    if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      let constructor = constructors.shift();
      return new this(
        properties,
        ...items.map(i => {
          return constructor ? constructor.from(i, ...constructors) : i;
        })
      );
    }
  }
}
