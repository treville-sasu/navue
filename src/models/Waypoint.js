import { Model } from "@/models/Base.js";
import { Angle } from "@/models/Quantities.js";

import { point, round } from "@turf/helpers";
import rhumbDistance from "@turf/rhumb-distance";
import rhumbBearing from "@turf/rhumb-bearing";
import rhumbDestination from "@turf/rhumb-destination";
import truncate from "@turf/truncate";
import greatCircle from "@turf/great-circle";

export class Waypoint extends Model {
  constructor([longitude, latitude, altitude] = [], properties = {}) {
    if (altitude !== undefined) {
      properties.altitudeReference || (properties.altitudeReference = "WGS84");
      altitude = round(Number(altitude));
    }
    super(
      point([round(longitude, 6), round(latitude, 6), altitude], properties)
    );
  }

  set longitude(val) {
    this.geometry.coordinates[0] = round(val, 6);
  }

  set latitude(val) {
    this.geometry.coordinates[1] = round(val, 6);
  }

  get altitude() {
    return this.geometry.coordinates[2];
  }

  set altitude(val) {
    this.geometry.coordinates[2] = round(val);
  }

  get lngLat() {
    return {
      lng: this.geometry.coordinates[0],
      lat: this.geometry.coordinates[1],
    };
  }

  set lngLat({ lng, lat }) {
    this.longitude = lng;
    this.latitude = lat;
  }

  distanceTo(wp) {
    return round(
      rhumbDistance(this.geometry.coordinates, wp.geometry.coordinates, {
        units: "meters",
      })
    );
  }

  bearingTo(wp) {
    return rhumbBearing(this.geometry.coordinates, wp.geometry.coordinates);
  }

  greatCircleTo(wp, properties) {
    let gc = truncate(
      greatCircle(this.geometry.coordinates, wp.geometry.coordinates, {
        npoints: 8,
        properties
      })
    );

    gc.properties = {
      ...gc.properties,
      distance: this.distanceTo(wp),
      bearing: this.bearingTo(wp)
    };

    return gc;
  }

  destinationPoint(distance, heading, properties) {
    const coords = rhumbDestination(
      this.geometry.coordinates,
      distance,
      Angle.wrap180(heading),
      {
        units: "meters",
      }
    ).geometry.coordinates;
    if (this.geometry.coordinates[2]) coords[2] = this.geometry.coordinates[2];
    return new this.constructor(coords, properties);
  }

  toGeoJSON(type) {
    switch (type) {
      case "Position":
        return this.geometry.coordinates;
      case "Point":
        return this.geometry;
      default:
        return {
          ...this,
          type: "Feature",
        };
    }
  }

  static from({
    type,
    geometry,
    coordinates,
    properties = {},
    ...otherProps
  } = {}) {
    if (arguments[0] instanceof this) return arguments[0];

    switch (type) {
      case this.name:
      case "Feature":
        if (geometry && geometry.type == "Point")
          coordinates = geometry.coordinates;
        else
          throw `Invalid data : 'geometry.type' should be 'Point' got '${
            geometry && geometry.type
          }'`;
        break;
      case "Point":
        // coordinates = coordinates;
        break;
      default:
        throw `Invalid data : 'type' should be 'Feature', 'Point' or '${this.name}' got '${type}'`;
    }
    return new this(coordinates, {
      ...otherProps,
      ...properties,
    });
  }

  static fromEvent({ lngLat: { lng, lat } } = {}) {
    return new this([lng, lat]);
  }
}
