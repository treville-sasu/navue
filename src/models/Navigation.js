import { Model, Collection } from "@/models/Base.js";
import { Waypoint } from "@/models/Waypoint.js";

export class Navigation extends Model {
  constructor({ name, notes, routes = [] } = {}) {
    routes = RouteCollection.from(routes);
    super({ name, notes, routes });
    this.type = "navigation";
  }

  addRoute(...args) {
    return this.routes.append(...args);
  }

  removeRoute(rte) {
    return rte instanceof Collection
      ? this.routes.remove(rte)
      : this.routes.splice(rte, 0);
  }

  addWaypoint({ insertBefore, ...wp } = {}, route = this.routes.last()) {
    return insertBefore ? route.insert(wp, insertBefore) : route.append(wp);
  }

  removeWaypoint(wp, route = this.routes.last()) {
    return wp instanceof Waypoint ? route.remove(wp) : route.splice(wp, 1);
  }

  getNextWaypoint(wp) {
    const wpId = this.routes.flat().indexOf(wp);
    return this.routes.flat()[wpId + 1];
  }
}

export class RouteCollection extends Collection {
  create(...args) {
    return Route.from(...args);
  }

  static from(collection) {
    const routeCollection = new RouteCollection();
    for (const route of collection) {
      if (Array.isArray(route)) routeCollection.append(route);
    }
    return routeCollection;
  }
}
export class Route extends Collection {
  create(...args) {
    return new Waypoint(...args);
  }

  static from(collection) {
    const route = new Route();
    if (Array.isArray(collection))
      for (const waypoint of collection) {
        route.append(waypoint);
      }
    return route;
  }
}
