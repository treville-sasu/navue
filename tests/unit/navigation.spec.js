import { RouteCollection, Route, Navigation } from "@/models/Navigation.js";
import { Waypoint } from "@/models/Waypoint.js";
import { Model, Collection } from "../../src/models/Base";

describe("routeCollection", () => {
  it("extends Collection", () => {
    expect(RouteCollection.prototype).toBeInstanceOf(Collection);
  });
  it("create RouteCollection with values", () => {
    // expect(new RouteCollection(1, 2)).toEqual([1, 2]);
    expect(new RouteCollection(1, 2)).toStrictEqual(
      expect.arrayContaining([1, 2])
    );
  });
  it("create a Route from, RouteCollection", () => {
    expect(new RouteCollection().create()).toBeInstanceOf(Route);
  });
});
describe("route", () => {
  it("extends Collection", () => {
    expect(Route.prototype).toBeInstanceOf(Collection);
  });
  it("create Route with values", () => {
    expect(new Route(1, 2)).toStrictEqual(expect.arrayContaining([1, 2]));
  });
  it("create a Waypoint from, Route", () => {
    expect(new Route().create()).toBeInstanceOf(Waypoint);
  });
});

describe("navigation", () => {
  it("extends Model", () => {
    expect(Navigation.prototype).toBeInstanceOf(Model);
  });
  it("create with defaults", () => {
    const nav = new Navigation();
    expect(nav).toHaveProperty("name", undefined);
    expect(nav).toHaveProperty("notes", undefined);
    expect(nav).toHaveProperty("routes", expect.any(RouteCollection));
  });

  it("create with properties & rejects unvalid property", () => {
    const nav = new Navigation({
      name: "test Nav",
      surname: "kiki",
      notes: "some text",
      routes: [
        [
          { name: "A", latitude: 0, longitude: 0 },
          { name: "B", latitude: 0, longitude: 1 }
        ],
        "some text at the wrong place"
      ]
    });
    expect(nav).toHaveProperty("name", "test Nav");
    expect(nav).toHaveProperty("notes", "some text");
    expect(nav).not.toHaveProperty("surname");
    expect(nav).toHaveProperty(
      "routes",
      new RouteCollection(
        new Route(
          new Waypoint({ name: "A", latitude: 0, longitude: 0 }),
          new Waypoint({ name: "B", latitude: 0, longitude: 1 })
        )
      )
      // new RouteCollection(new Route("someObject"))
      // expect.arrayContaining(new Route("someObject"))
    );
  });

  it("add a new route", () => {
    const nav = new Navigation();
    expect(nav.addRoute()).toBeInstanceOf(Route);
    expect(nav.routes).toHaveLength(1);
    expect(nav.routes).toEqual(new RouteCollection(new Route()));
  });

  it("remove a given route", () => {
    const navigation = new Navigation();
    const route = navigation.addRoute();
    navigation.removeRoute(route);
    expect(navigation.routes).toHaveLength(0);
  });
});
