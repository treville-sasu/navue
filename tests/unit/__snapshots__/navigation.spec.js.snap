// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation export toGeoJSON as Great Circle 1`] = `
Object {
  "bbox": Array [
    0,
    0,
    3,
    4,
  ],
  "features": Array [
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            0,
            0,
            10,
          ],
          Array [
            1,
            1,
            20,
          ],
        ],
        "type": "LineString",
      },
      "id": 1,
      "properties": Object {
        "altitudeReference": "WGS84",
        "name": "A",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            1,
            1,
            20,
          ],
          Array [
            1,
            2,
            20,
          ],
        ],
        "type": "LineString",
      },
      "id": 2,
      "properties": Object {
        "altitudeReference": "WGS84",
        "name": "Babababa",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            2,
            2,
            20,
          ],
          Array [
            3,
            3,
            10,
          ],
        ],
        "type": "LineString",
      },
      "id": 1,
      "properties": Object {
        "altitudeReference": "WGS84",
        "name": "C",
      },
      "type": "Feature",
    },
    Object {
      "geometry": Object {
        "coordinates": Array [
          Array [
            3,
            3,
            10,
          ],
          Array [
            3,
            4,
            10,
          ],
        ],
        "type": "LineString",
      },
      "id": 2,
      "properties": Object {
        "altitudeReference": "WGS84",
        "name": "Dadadadada",
      },
      "type": "Feature",
    },
  ],
  "type": "FeatureCollection",
}
`;

exports[`navigation hold a list of POI 1`] = `
Array [
  "LFBA",
  "LFDE",
]
`;
